### PR TITLE
Introduce tools.codegen.api.translate

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49505 Add at::cpu namespace of functions for structured kernels
* #49498 Push anonymous namespace into codegen, not template
* #49490 Make use_c10_dispatcher: full mandatory for structured kernels
* #49489 Move default or no default logic into native.argument
* #49348 codegen: Resolve overload ambiguities created by defaulted arguments
* #49341 Tighten up error checking on manual_kernel_registration
* #49245 Construct CppSignatureGroup from NativeFunction
* **#49122 Introduce tools.codegen.api.translate**

cpparguments_exprs has induced a lot of head scratching in many recent PRs for how to structure the code in a good way.  This PR eliminates the old algorithm for an entirely new algorithm inspired by logic programming.  The net result is shorter, cleaner and should be more robust to future changes.

This PR is a bit of a whopper.  Here is the order to review it.

- tools/codegen/api/types.py
  - Deleted CppArgument, CppArgumentPackIface (and subclasses), CppExpr, DispatcherExpr, DispatcherArgument, NativeExpr, NativeArgument, MetaArgument. All things previously called XArgument are now Binding. All things previously called XExpr are now Expr. I deleted the `__str__` implementation on Binding and fixed all call sites not to use it. On Binding, I renamed `str_no_default` and `str_default` to `defn` and `decl` for better symmetry with the corresponding signature concepts, although I'm open to naming them back to their original versions.
  - Obviously, things are less type safe without the class distinctions. So I introduce a new ADT called CType. CType represents the *semantic C++ type* of a binding: it is both the C++ type (e.g., `const Tensor&`) as well as the argument name that specifies what the  binding denotes (e.g., `other`). Every binding now records its CType. The key observation here is that you don't actually care if a given expression is from the cpp or dispatcher or native API; what you care is having enough information to know what the expression means, so you can use it appropriately. CType has this information. For the most part, ArgNames are just the string names of the arguments as you see them in JIT schema, but there is one case (`possibly_redundant_memory_format`) where we encode a little extra information. Unlike the plain strings we previously used to represent C++ types, CType have a little bit of structure around optional and references, because the translation code needs to work around these concepts.
  - I took the opportunity to kill all of the private fields like `_arguments` and `_returns_type` (since the argument types don't make sense anymore). Everything is computed for you on the fly. If this is a perf problem in codegen we can start using `cached_property` decorator.
  - All of the heavy lifting in CppSignature.argument_packs has been moved to the cpp module. We'll head over there next. Similarly, all of the exprs methods are now calling translate, the new functionality which we haven't gotten to yet
- tools/codegen/api/cpp.py
   - We refactor all of the type computation functions to return CType instead of str. Because CTypes need to know the denotation, there is a new `binds: ArgName` argument to most functions that provides the denotation, so we can slot it in. (An alternative would have been to construct CTypes without denotations and then fill them in post-facto, but I didn't do it this way. One downside is there are some places where I need a CType without denotation, so I fill these in with `__placeholder__` whenever this happens).
  - `argument` and `arguments` are now extremely simple. There is no more Pack business, just produce one or more Bindings. The one thing of note is that when both a `memory_format` and `options` are in scope, we label the memory format as `possibly_redundant_memory_format`. This will be used in translation
- tools/codegen/api/dispatcher.py and tools/codegen/api/native.py - same deal as cpp.py. One thing is that `cpparguments_exprs` is deleted; that is in the translator
- tools/codegen/api/translate.py - the translator! It uses a very simple backwards deduction engine to work out how to fill in the arguments of functions. There are comments in the file that explain how it works.
- Everything else: just some small call site tweaks for places when I changed API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25455887](https://our.internmc.facebook.com/intern/diff/D25455887)

### CODE REVIEW

File: tools/codegen/api/cpp.py [hunk 1]
```diff
@@ -36,90 +36,96 @@ def name(func: FunctionSchema, *, faithful_name_for_out_overloads: bool = False)
...
-            return f"std::array<bool,{t.size}>"
+            return BaseCType(f"std::array<bool,{t.size}>", binds)
         else:
             return None
     else:
         raise AssertionError(f"unrecognized type {repr(t)}")
 
 # Translation of types occuring in JIT arguments to a C++ argument type.
-def argumenttype_type(t: Type, *, mutable: bool) -> str:
+def argumenttype_type(t: Type, *, mutable: bool, binds: ArgName) -> CType:
```
Comment (peterbell10) [inline tools/codegen/api/cpp.py:72]: >An alternative would have been to construct CTypes without denotations and then fill them in post-facto, but I didn't do it this way. One downside is there are some places where I need a CType without denotation, so I fill these in with `__placeholder__` whenever this happens

How about returning a structured object that represents only the type without semantic information (e.g. `RawCType`). And `CType` could just be a simple data class:
```python
RawCType = Union[BaseRawCType, OptionalRawCType, ConstRefRawCType, MutRefRawCType]

@dataclass(frozen=True)
class CType:
    type: RawCppType
    name: ArgName
```

`argumenttype_type` can return a `RawCType` so it doesn't need to pass the `binds` argument everywhere. You also never need to pass `__placeholder__`. i.e. instead of 
```python
semantic_type = argument_type(t, binds=name)
str_type = argument_type(t, binds='__placeholder__').cpp_type()
```
you would have
```python
semantic_type = CType(type=argument_type(t), name=name)
string_type = argument_type(t).str()
```

Personally, I probably also rename `CType` to something else like `SemType` or `CArgument` so that `RawCType` could just be `CType`/`CppType`.

Comment (ezyang) [reply tools/codegen/api/cpp.py:72]: The trouble with this is that it makes some operations on arguments require more unwrapping/wrapping. Imagine that you have `CType(ConstRefRawCType(BaseRawCType('Tensor')), 'other')`. A semantics preserving transformation would be to remove `ConstRefRawCType`, giving `CType(BaseRawCType('Tensor'), 'other')`. But I can't just say `orig_ctype.elem` as I did before; here I have to do `CType(orig_ctype.type.elem, orig_ctype.name)`.

Comment (ezyang) [reply tools/codegen/api/cpp.py:72]: Also, seriously considered using SemType, but it's more characters XD XD XD

Comment (peterbell10) [reply tools/codegen/api/cpp.py:72]: If I'm reading this right, the actual `SemType` unwrapping is only done in one place (line 100 of transform.py). All other cases actually only care about the (non-semantic) C++ type information.

Either way, If it does come up a lot it could be made terser with some short helpers:
```python
def unwrap(self: SemType) -> SemType:
    return SemType(self.type.elem, self.name)

def wrap(self: SemType, kls: Callable) -> SemType:
    return SemType(kls(self.type), self.name)
```

Comment (ezyang) [reply tools/codegen/api/cpp.py:72]: > If I'm reading this right, the actual SemType unwrapping is only done in one place (line 100 of transform.py). All other cases actually only care about the (non-semantic) type information.

That may very well be true how the code is today. There's some deeper aesthetic reason why I really want the ArgName on the leaves is because it also generalizes correctly when you have tuples: `PairCType(BaseRawCType('Tensor', 'one'), BaseRawCType('Tensor', 'two'))`. You have to do a parallel tree for annotations to represent this correctly. By the way, this is exactly how alias annotations were added to JIT schema type system (out-of-line) and I hate it haha.

By the way, the helper functions aren't well typed. You don't statically know that SemType's self.type is a class with an elem object. (And wrap is just ick!)

File: tools/codegen/api/translate.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            memory_format = direct_solve(
+                OptionalCType(BaseCType("MemoryFormat", SpecialArgName.possibly_redundant_memory_format))
+            )
+            try:
+                options = direct_solve(options_ctype)
+                return f"c10::impl::check_tensor_options_and_extract_memory_format({options}, {memory_format})"
+            except UnsatError:
+                return memory_format
+
+        elif goal == BaseCType("TensorOptions", "options"):
```
Comment (ljk53) [inline tools/codegen/api/translate.py:118]: I haven't verified all the details yet, but the idea looks good me.

The python codegen did similar stuff to some extent (pulling "goal" from available "bindings" in the context): https://github.com/pytorch/pytorch/pull/46244#discussion_r506575407 :p

(note to myself) One natural idea is to apply the translator to python codegen and autograd codegen as well, but the autograd codegen often emits extra initialization statements, and TensorOptions definition there is slightly different (has requires_grad field). Maybe we will need a different translator for the "top-half" of the codegen.

**Review by ljk53 — APPROVED** [review]
LGTM (assuming codegen output doesn't change)

File: tools/codegen/api/translate.py (hunk 2 above)
Comment (ezyang) [reply tools/codegen/api/translate.py:118]: Yeah, I'm not sure how possible it is. Need to read the code more carefully to tell!

File: tools/codegen/api/translate.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+# "pin_memory"; you can't just use a random other boolean in a context where you
+# need a "pin_memory"!
+#
+# The translator takes a list of needed CTypes, and then figures out how
+# to construct expressions with these CTypes from the given bindings.  Many
+# of these expressions are trivial (I need a Tensor other; there's a Tensor
+# other scope); others are more nontrivial and may require packing/unpacking.
+# Some examples of non-trivial action:
+#
+#   - Need the "dtype" binding?  Well, maybe "dtype" isn't available
```
Comment (smessmer) [inline tools/codegen/api/translate.py:20]: Gathering isn't be necessary anymore since all ops with TensorOptions are already c10-full. You should only need scattering in some situations

File: tools/codegen/api/types.py [hunk 4]
```diff
@@ -1,145 +1,116 @@
...
+# just use a random other boolean in a context where you need a "pin_memory"!
+#
+# CTypes encode C++ type structure as needed for translation.  Right now we
+# track references and optional, but don't, for example, track ArrayRef.  If
+# you need trnsnlations that know about these types, beef up this data
+# structure.
 
-#                           cpp types
+@dataclass(frozen=True)
+class BaseCType:
```
Comment (smessmer) [inline tools/codegen/api/types.py:30]: I find the term "semantic type" confusing, I liked "argument" better. Or actually, a type plus a name is a "parameter" rather.

**Review by smessmer — COMMENTED** [review]
It generally looks like this is doing what you claim, so I'm not strictly opposed to it, but I find it a bit complex. I guess that some of the complexity is coming from the worthy goal of bringing order into the weird situation we had before where we needed to match n API styles to n dispatcher styles. (n=2, it was 2 API styles and 2 dispatcher styles). Mapping n:n is complex and logic programming allows for a reasonable abstraction layer to do it. But actually, once everything is c10 full, there is only one dispatcher style. Mapping n:1 can be done with much easier code. And basically, all the cases that matter are already c10 full, for example the code path to gather arguments is likely dead code already. Note that still `n=2`, so it's 2:1, but even if we add more APIs later, n:1 is still simple enough.

File: tools/codegen/api/types.py (hunk 4 above)
Comment (ezyang) [reply tools/codegen/api/types.py:30]: The bigest trouble with the name Argument is that CTypes also accurately classify C++ expressions, which are definitely not arguments. (Although, I'm a little confused where you attached this comment: Cpp/DispatcherArgument got renamed to Binding, but I'm not opposed to renaming Binding back to Argument, except that conflicts with model.Argument which already took this name. CArgument would be in line with CType naming convention.

File: tools/codegen/api/translate.py (hunk 3 above)
Comment (ezyang) [reply tools/codegen/api/translate.py:20]: I'll try deleting the gather rule and see if it works.

**Comment by ezyang** [comment]
> Mapping n:1 can be done with much easier code. And basically, all the cases that matter are already c10 full, for example the code path to gather arguments is likely dead code already. Note that still n=2, so it's 2:1, but even if we add more APIs later, n:1 is still simple enough.

This is a legitimate point. However, I kind of feel like n:n wasn't the tall pole in the complexity tent (since we only really had two translators even then); the big problem was gather (which, by the way, I've checked, and it indeed is not being used anymore after your patchset--so your broader point still stands).

That being said, I think having n:n working can open up new use cases. For example, here https://github.com/pytorch/pytorch/pull/49505/files#diff-6382c847dac2d011f822c492e61f9fa0eceb47c2d6fc5fb4f0c42af726c1a3aaR467 I used the translator API to translate from cpp API to native API, and it worked straight out of the box. Of course, part of the reason it worked straight out of the box is because native is the same as dispatcher for c10_dispatcher_full operators, but it is easy for me to trust that the translation will fail when these don't coincide, because it is entirely type-directed.
